File: src/com/rian/difficultycalculator/skills/standard/StandardFlashlight.java
package com.rian.difficultycalculator.skills.standard;

import com.rian.difficultycalculator.beatmap.hitobject.DifficultyHitObject;
import com.rian.difficultycalculator.evaluators.standard.StandardFlashlightEvaluator;

import java.util.EnumSet;

import main.osu.game.mods.GameMod;

/**
 * Represents the skill required to memorize and hit every object in a beatmap with the Flashlight mod enabled.
 */
public class StandardFlashlight extends StandardSkill {
    private double currentStrain;
    private final boolean hasHidden;

    /**
     * @param mods The mods that this skill processes.
     */
    public StandardFlashlight(EnumSet<GameMod> mods) {
        super(mods);

        hasHidden = mods.contains(GameMod.MOD_HIDDEN);
    }

    @Override
    protected double strainValueAt(DifficultyHitObject current) {
        currentStrain *= strainDecay(current.deltaTime);
        double skillMultiplier = 0.052;
        currentStrain += StandardFlashlightEvaluator.evaluateDifficultyOf(current, hasHidden) * skillMultiplier;

        return currentStrain;
    }

    @Override
    protected double calculateInitialStrain(double time, DifficultyHitObject current) {
        return currentStrain * strainDecay(time - current.previous(0).startTime);
    }

    @Override
    protected void saveToHitObject(DifficultyHitObject current) {
        current.flashlightStrainWithSliders = currentStrain;
    }

    @Override
    protected int getReducedSectionCount() {
        return 0;
    }

    @Override
    protected double getReducedSectionBaseline() {
        return 1;
    }

    @Override
    protected double getDecayWeight() {
        return 1;
    }

    private double strainDecay(double ms) {
        return Math.pow(0.15, ms / 1000);
    }
}

File: src/com/rian/osu/difficulty/skills/StandardFlashlight.kt
package com.rian.osu.difficulty.skills

import com.rian.osu.difficulty.StandardDifficultyHitObject
import com.rian.osu.difficulty.evaluators.StandardFlashlightEvaluator.evaluateDifficultyOf
import com.rian.osu.mods.Mod
import kotlin.math.pow

/**
 * Represents the skill required to memorize and hit every object in a beatmap with the Flashlight mod enabled.
 */
class StandardFlashlight(
    /**
     * The [Mod]s that this skill processes.
     */
    mods: Iterable<Mod>
) : StandardStrainSkill(mods) {
    override val reducedSectionCount = 0
    override val reducedSectionBaseline = 1.0
    override val decayWeight = 1.0

    private var currentStrain = 0.0
    private val skillMultiplier = 0.05512
    private val strainDecayBase = 0.15

    override fun strainValueAt(current: StandardDifficultyHitObject): Double {
        currentStrain *= strainDecay(current.deltaTime)
        currentStrain += evaluateDifficultyOf(current, mods) * skillMultiplier

        return currentStrain
    }

    override fun calculateInitialStrain(time: Double, current: StandardDifficultyHitObject) =
        currentStrain * strainDecay(time - current.previous(0)!!.startTime)

    override fun difficultyValue() = currentStrainPeaks.sum()

    private fun strainDecay(ms: Double) = strainDecayBase.pow(ms / 1000)
}
